### ISSUE TITLE
Proposal: MonadError.transformWith

### ISSUE THREAD

**Issue by alexandru** [typelevel/cats#2161]
So `MonadError` has `attempt` and `handleErrorWith` and for many use-cases one has to introduce both `map` / `flatMap` transformations and `handleErrorWith` handlers in the same "frame".

E.g. what happens for data types implementing `cats-effect` like `cats.effect.IO` and `monix.eval.Task` is something like this

```scala
fa.attempt.flatMap {
  case Right(next) => continueLoop(next)
  case Left(error) => stopLoop(error)
}
```

This is more or less the equivalent of this pattern:

```scala
var streamError = true
try {
  val next = fa()
  streamError = false
  continueLoop(next)
} catch {
  case error if streamError => 
    stopLoop(error)
}
```

Well the problem is that usage of either `attempt` or `handleErrorWith` involves 2 `flatMap` (bind) frames instead of one. And as we all know, the performance of `IO` / `Task` data types, compared with plain function calls, is terrible.

The proposal is to introduce an alternative to `attempt.flatMap` like so:

```scala
def transformWith[A, B](fa: F[A])(bind: A => F[B], recover: Throwable => F[B]): F[B]
```

Just did.a fast [benchmark](https://gist.github.com/alexandru/603e8003fa8da45d5d37f58bf928cb0f) to show what the difference can be:

```
attempt             10000  thrpt   10  2383.878 ±  46.546  ops/s
onErrorHandleWith   10000  thrpt   10  2813.948 ± 176.036  ops/s
transformWith       10000  thrpt   10  4115.213 ±  40.821  ops/s
```

Precedents:

1. [Task.transformWith](https://monix.io/api/2.3/monix/eval/Task.html#transformWith[R](fa:A=%3Emonix.eval.Task[R],fe:Throwable=%3Emonix.eval.Task[R]):monix.eval.Task[R])
2. [Future.transformWith](http://www.scala-lang.org/api/2.12.3/scala/concurrent/Future.html#transformWith[S](f:scala.util.Try[T]=%3Escala.concurrent.Future[S])(implicitexecutor:scala.concurrent.ExecutionContext):scala.concurrent.Future[S])

Note — I'd prefer the version with the two function parameters, because I'd rather avoid the extra `Either` boxing. 

Implementation notes:

1. `cats.effect.IO` is already optimized for this operation
2. `monix.eval.Task` is already optimized for this operation

Basically I'd like to not feel bad about adding error handling.

**Comment by joan38** [comment]
Isn't this solved by `redeem`?

**Comment by opensorceror** [comment]
@joan38 It doesn't look like `redeem` is not available on `MonadError`. (It _is_ available, however, on the `cats.effect.IO` type.)

**Comment by joan38** [comment]
@opensorceror see https://github.com/typelevel/cats/pull/2237#issuecomment-508542321
